hazelcast/src/main/java/com/hazelcast/internal/diagnostics/EventQueuePlugin.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.diagnostics;
+
+import com.hazelcast.cache.impl.CacheEventData;
+import com.hazelcast.cache.impl.CacheEventSet;
+import com.hazelcast.collection.impl.collection.CollectionEvent;
+import com.hazelcast.collection.impl.list.ListService;
+import com.hazelcast.collection.impl.queue.QueueEvent;
+import com.hazelcast.collection.impl.set.SetService;
+import com.hazelcast.core.EntryEventType;
+import com.hazelcast.map.impl.event.EntryEventData;
+import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl;
+import com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher;
+import com.hazelcast.spi.properties.HazelcastProperties;
+import com.hazelcast.spi.properties.HazelcastProperty;
+import com.hazelcast.util.ItemCounter;
+import com.hazelcast.util.executor.StripedExecutor;
+
+import java.text.NumberFormat;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+
+import static com.hazelcast.internal.diagnostics.Diagnostics.PREFIX;
+import static java.lang.Math.min;
+import static java.lang.String.format;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * The EventQueuePlugin checks the event queue and samples the event types if the size is above a certain threshold.
+ * This is very useful to figure out why the event queue is running full.
+ */
+public class EventQueuePlugin extends DiagnosticsPlugin {
+
+    /**
+     * The period in seconds this plugin runs.
+     *
+     * With the EventQueuePlugin one can see what is going on inside the event queue.
+     * It makes use of sampling to give some impression of the content.
+     *
+     * If set to 0, the plugin is disabled.
+     */
+    public static final HazelcastProperty PERIOD_SECONDS
+            = new HazelcastProperty(PREFIX + ".event.queue.period.seconds", 0, SECONDS);
```

Comment:
It doesn't print an uncontrolled number of things. I copied the plugin from the `OverloadedConnectionsPlugin`, which does sampling. Nevertheless the sampling can be costly, since a lot of strings are built, to show the content of the events. Exactly the same as with the `OverloadedConnectionsPlugin` and the packet content. Super useful, but more like an on-demand plugin.